Update to UFS as of 2021 Oct 07

WalterKolczynski-NOAA opened the pull request:
This update adds several new stochastic physics possibilities, with
associated switched added. There are also some other new namelist
settings.

MOM input template files have been updated to use the @[ ] notation
following same changes in UFS, with corresponding changes to the
replacement code. The new template files were copied from UFS, except
saving of initial conditions is turned on.

Due to changes in how the output cadence is defined in the namelist,
an explicit list of forecast hours is generated in the pre-det script.
To avoid duplication, this list is used later to create the symlinks
for the output files.

WalterKolczynski-NOAA committed:
Update to UFS version as of 2021 Oct 07

This update adds several new stochastic physics possibilities, with
associated switched added. There are also some other new namelist
settings.

MOM input template files have been updated to use the @[ ] notation
following same changes in UFS, with corresponding changes to the
replacement code. The new template files were copied from UFS, except
saving of initial conditions is turned on.

Due to changes in how the output cadence is defined in the namelist,
an explicit list of forecast hours is generated in the pre-det script.
To avoid duplication, this list is used later to create the symlinks
for the output files.

WalterKolczynski-NOAA commented:
Both a full coupled and an atm-only, forecast-only test ran to completion on Orion.

KateFriedman-NOAA commented:
@WalterKolczynski-NOAA Need to remove duplicate if-block in build_all.sh:

https://github.com/WalterKolczynski-NOAA/global-workflow/blob/feature/update_ufs/sorc/build_all.sh#L61:L71

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA Need to remove duplicate if-block in build_all.sh:
> 
> https://github.com/WalterKolczynski-NOAA/global-workflow/blob/feature/update_ufs/sorc/build_all.sh#L61:L71

I didn't touch build_all this PR. I'll fix it after.

KateFriedman-NOAA reviewed (approved):
Looks good but I'll need others to review the science aspects.

JessicaMeixner-NOAA commented:
Two questions as I'm reviewing: 
1. The updating of the Mom input is great! Would this be a good time to just point to the ufs-weather-model/test/parm/MOM* input files as well?  Which could help simplify updates in the future? 
2.  Are you updating the nameless based on a specific rt test in ufs-weather-model? And if so, which one?

WalterKolczynski-NOAA commented:
> Two questions as I'm reviewing:
> 
> 1. The updating of the Mom input is great! Would this be a good time to just point to the ufs-weather-model/test/parm/MOM* input files as well?  Which could help simplify updates in the future?

We talked about this on the call today, and we're looking towards that. Right now, I made one manual change for saving the initial files, which is hardcoded to false in the RTs.

> 2. Are you updating the namelist based on a specific rt test in ufs-weather-model? And if so, which one?

I used defaults from the coupled control as much as possible, though I leave the new options (like ocean perturbation) off for now. So the default settings for all the new SP pieces are there, but they default to off. Turning them on will bring all the default settings with it, unless they are overridden.

WalterKolczynski-NOAA commented:
Basically, I made hooks for everything but kept things as much the same as before as possible. But for the options that have a bunch of settings that come along when they are turned on, all the defaults from the coupled RT are sitting there ready for when they are turned on.

JessicaMeixner-NOAA commented on ush/parsing_namelists_CICE.sh:
```diff
@@ -105,7 +105,7 @@ cat > ice_in <<eof
    ndte            = 120
    revised_evp     = .false.
-   kevp_kernel     = 0
+   evp_algorithm   = 'standard_2d'
```
&setup_nml
sec_init = @[SECS]
was added in tests/parm/ice_in_template but not here.

WalterKolczynski-NOAA replied:
> Looking at changes that were updated in ufs-weahter-model:  [ufs-community/ufs-weather-model@b26a896...develop](https://github.com/ufs-community/ufs-weather-model/compare/b26a896f2c9bada438414a51218bc72236925b8c...develop) I saw that ufs-weather-model modules are updated to use hpc stack 1.2.0 instead of 1.1.0 and that change is not reflected in the module files here.  I'm not sure if that's intentional or not?  In addition, the ESMF and FMS modules have been updated: module load esmf/8_2_0_beta_snapshot_14 module load fms/2021.03which is also not reflected in the module files in this update. 

The hpc-stack update seems to only be on wcoss2.

For esmf and fms, I'm not sure if it currently even matters, since it is correct at compile time (do we even need to be loading these at runtime?). But I will update them for consistency. In the slightly wider view: with the new dictates from NCO, module updates for ALL workflow components need to start being coordinated because NCO wants a single top-level versions file to control all modules.

> 
> Someone else would be a better person to review the updates related to the output frequency.
> 
> I'm still reviewing the namelist changes, do you have output from your run or should I just make a new run myself? I can tell the changes for MOM6 look okay, I've made one comment on CICE, but for input.nml - I'm still following all the diffs and it might be easier if there's output to just look at instead of just looking at the code.

/work/noaa/global/wckolczyn/noscrub/global-workflow/update_ufs
/work/noaa/global/wckolczyn/noscrub/global-workflow/update_ufs_atm

If you need the run directories, you'll have to rerun it, as I neglected to change KEEPDATA. I'm still following the directions I listed in PR #430 to run the coupled runs.

WalterKolczynski-NOAA replied:
Actually, I'm going to rerun at least the coupled with these new changes, and I'm turning KEEPDATA on if you want to wait for that.

WalterKolczynski-NOAA replied:
The run directory will be in /work/noaa/stmp/wkolczyn/RUNDIRS/update_ufs/2013040100/gfs when it is created.

I had to scrub the output from the rotdir, so that will be regenerated as well.

JessicaMeixner-NOAA replied:
Thanks Walter! I will look at your new output.

JessicaMeixner-NOAA reviewed (changes requested):
Looking at changes that were updated in ufs-weahter-model: 
https://github.com/ufs-community/ufs-weather-model/compare/b26a896f2c9bada438414a51218bc72236925b8c...develop
I saw that ufs-weather-model modules are updated to use hpc stack 1.2.0 instead of 1.1.0 and that change is not reflected in the module files here.  I'm not sure if that's intentional or not?  In addition, the ESMF and FMS modules have been updated: module load esmf/8_2_0_beta_snapshot_14
module load fms/2021.03which is also not reflected in the module files in this update. 


Someone else would be a better person to review the updates related to the output frequency.    

I'm still reviewing the namelist changes, do you have output from your run or should I just make a new run myself?  I can tell the changes for MOM6 look okay, I've made one comment on CICE, but for input.nml - I'm still following all the diffs and it might be easier if there's output to just look at instead of just looking at the code.

JessicaMeixner-NOAA commented:
Thanks for sharing the output.  From what I saw comparing input.nml files with regression tests, we could add 
&gfs_physics_nml
use_cice_alb = @[USE_CICE_ALB]
which is currently set to its default value false, but this could always be added in a separate PR.

WalterKolczynski-NOAA committed:
Update ESMF & FMS and add new CICE setting

Updates the ESMF and FMS to match the more recent modules used by UFS.

Added new sec_init to CICE namelist.

WalterKolczynski-NOAA commented:
Coupled run with the new changes completed successfully. Due to some configuration change on Orion, I had to add `unset SLURM_MEM_PER_CPU` to the end of config.base in order for srun to work. I have an email into the Orion helpdesk on this issue.

JessicaMeixner-NOAA reviewed (approved):
Thanks for the updates @WalterKolczynski-NOAA!
